Framework/Geometry/src/Instrument/ComponentInfo.cpp

```diff
@@ -107,8 +107,39 @@ ComponentInfo::componentsInSubtree(size_t componentIndex) const {
   return m_componentInfo->componentsInSubtree(componentIndex);
 }
 
+const std::vector<size_t> &
+ComponentInfo::children(size_t componentIndex) const {
+  return m_componentInfo->children(componentIndex);
+}
+
 size_t ComponentInfo::size() const { return m_componentInfo->size(); }
 
+ComponentInfo::QuadrilateralComponent
+ComponentInfo::quadrilateralComponent(const size_t componentIndex) const {
+  auto type = componentType(componentIndex);
+  if (!(type == Beamline::ComponentType::Structured ||
+        type == Beamline::ComponentType::Rectangular))
+    throw std::runtime_error("ComponentType is not Structured or Rectangular "
+                             "in ComponentInfo::quadrilateralComponent.");
+
+  QuadrilateralComponent corners;
+  auto innerRangeComp =
+      m_componentInfo->componentRangeInSubtree(componentIndex);
+  // nSubComponents, subtract off self hence -1. nSubComponents = number of
+  // horizontal columns.
+  corners.nX = innerRangeComp.end() - innerRangeComp.begin() - 1;
+  auto innerRangeDet = m_componentInfo->detectorRangeInSubtree(componentIndex);
+  auto nSubDetectors =
+      std::distance(innerRangeDet.begin(), innerRangeDet.end());
+  corners.nY = nSubDetectors / corners.nX;
+
+  corners.bottomLeft = *innerRangeDet.begin();
+  corners.topRight = corners.bottomLeft + nSubDetectors - 1;
+  corners.topLeft = corners.bottomLeft + (corners.nY - 1);
+  corners.bottomRight = corners.topRight - (corners.nY - 1);
+  return corners;
```

Comment:
Didn't we decide that this can give the wrong results for Rectangular banks where the id ordering is by row? I think it would be advisable to bring those changes across. Potentially, even to push a branch to master as part of another PR, as it's technically a bug affecting the upcoming release.

Reply:
This is changed in the next PR with the optimisations [here](https://github.com/mantidproject/mantid/pull/21880)